spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java

```diff
@@ -0,0 +1,239 @@
+  }
+
+  private Dataset<Row> buildActualFileDF() {
+    List<String> topLevelDirs = Lists.newArrayList();
+    List<String> matchingTopLevelFiles = Lists.newArrayList();
+
+    try {
+      Path path = new Path(location);
+      FileSystem fs = path.getFileSystem(hadoopConf.value());
+
+      for (FileStatus file : fs.listStatus(path, HiddenPathFilter.get())) {
+        if (file.isDirectory()) {
+          topLevelDirs.add(file.getPath().toString());
+        } else if (file.isFile() && file.getModificationTime() < olderThanTimestamp) {
+          matchingTopLevelFiles.add(file.getPath().toString());
+        }
+      }
+    } catch (IOException e) {
+      throw new RuntimeIOException(e, "Failed to determine top-level files and dirs in {}", location);
+    }
+
+    JavaRDD<String> matchingTopLevelFileRDD = sparkContext.parallelize(matchingTopLevelFiles, 1);
+
+    if (topLevelDirs.isEmpty()) {
+      return spark.createDataset(matchingTopLevelFileRDD.rdd(), Encoders.STRING()).toDF("file_path");
+    }
+
+    int parallelism = Math.min(topLevelDirs.size(), partitionDiscoveryParallelism);
+    JavaRDD<String> topLevelDirRDD = sparkContext.parallelize(topLevelDirs, parallelism);
+
+    Broadcast<SerializableConfiguration> conf = sparkContext.broadcast(hadoopConf);
+    JavaRDD<String> matchingLeafFileRDD = topLevelDirRDD.mapPartitions(listDirsRecursively(conf, olderThanTimestamp));
+
+    JavaRDD<String> matchingFileRDD = matchingTopLevelFileRDD.union(matchingLeafFileRDD);
+    return spark.createDataset(matchingFileRDD.rdd(), Encoders.STRING()).toDF("file_path");
+  }
+
+  private static FlatMapFunction<Iterator<String>, String> listDirsRecursively(
+      Broadcast<SerializableConfiguration> conf,
+      long olderThanTimestamp) {
+
+    return (FlatMapFunction<Iterator<String>, String>) dirs -> {
+      List<String> files = Lists.newArrayList();
+      Predicate<FileStatus> predicate = file -> file.getModificationTime() < olderThanTimestamp;
+      dirs.forEachRemaining(dir -> {
+        List<String> dirFiles = listDirRecursively(dir, predicate, conf.value().value());
+        files.addAll(dirFiles);
+      });
+      return files.iterator();
+    };
+  }
+
+  private static List<String> listDirRecursively(String dir, Predicate<FileStatus> predicate, Configuration conf) {
+    try {
+      Path path = new Path(dir);
+      FileSystem fs = path.getFileSystem(conf);
+
+      List<String> childDirs = Lists.newArrayList();
+      List<String> matchingFiles = Lists.newArrayList();
+
+      for (FileStatus file : fs.listStatus(path, HiddenPathFilter.get())) {
+        if (file.isDirectory()) {
+          childDirs.add(file.getPath().toString());
+        } else if (file.isFile() && predicate.test(file)) {
+          matchingFiles.add(file.getPath().toString());
+        }
+      }
+
+      for (String childDir : childDirs) {
+        List<String> childDirFiles = listDirRecursively(childDir, predicate, conf);
```

Comment:
What about passing a result list into this method to avoid creating lots of small lists and merging them together in every call?

Reply:
Done.